File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/accuracy_vs_precision_tutorial.m
% Clean up
clear; clc; close all;

% Create a figure sized for clarity
figure('Position',[100 100 1200 600]);

%------------------------------------------------
% Common parameters for circles
%------------------------------------------------
radii = [1,2,3,4,5];    % Radii of concentric circles
angles_for_x = 0;     % Angle(s) where 'X' is placed (0 = top)
nCirclePoints = 200;  % Number of points to plot each circle

%------------------------------------------------
% Subplot (a): top - circles, bottom - distribution
%------------------------------------------------
% -- Top row (a)
subplot(2,4,1); 
hold on; axis equal;
title('(a)');
% Plot concentric circles centered at (0,0)
for r = radii
    [xc, yc] = circle(0,0,r,nCirclePoints);
    plot(xc, yc, 'k'); 
    % % Place an 'X' at specified angle(s)
    % for theta = angles_for_x
    %     xX = r*cosd(theta);
    %     yX = r*sind(theta);
    %     plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
    % end
end
xX = 1.5*cosd(0); yX = 1.5*sind(0);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 4.5*cosd(80); yX = 4.5*sind(80);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 4.5*cosd(150); yX = 4.5*sind(150);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 2.5*cosd(200); yX = 2.5*sind(200);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 4.5*cosd(300); yX = 4.5*sind(300);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);

% Plot dashed lines
xline(0, 'r--','LineWidth',1.5);  % Red dashed line at x=0
xline(0.75, 'g--','LineWidth',1.5);  % Green dashed line at x=2
xlim([-5 5]); ylim([-5 5]);
axis square
xlabel('X'); ylabel('Y');
set(gca,'Box','on');

% -- Bottom row (a)
subplot(2,4,5);
hold on;
% Create an example distribution
xvals = linspace(-10,10,200);
mu_a = 0.75;   % mean
sigma_a = 2; % std
yvals_a = normpdf(xvals, mu_a, sigma_a);
plot(xvals, yvals_a, 'k', 'LineWidth',1.5);
% Vertical dashed lines
xline(0,'r--','LineWidth',1.5);
xline(mu_a,'g--','LineWidth',1.5);
title('(a) Distribution');
xlabel('x'); ylabel('PDF');
xlim([-5 5]);
axis square
grid on;

%------------------------------------------------
% Subplot (b): top - circles offset, bottom - distribution
%   (You can change centers, lines, etc. to match your figure)
%------------------------------------------------
% -- Top row (b)
subplot(2,4,2); 
hold on; axis equal;
title('(b)');
% % Suppose we shift the center of the circles slightly, e.g. center=(1,0)
% centerB = [1, 0];
% for r = radii
%     [xc, yc] = circle(centerB(1), centerB(2), r, nCirclePoints);
%     plot(xc, yc, 'k');
%     % 'X' at top of each circle
%     for theta = angles_for_x
%         xX = centerB(1) + r*cosd(theta);
%         yX = centerB(2) + r*sind(theta);
%         plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
%     end
% end
for r = radii
    [xc, yc] = circle(0,0,r,nCirclePoints);
    plot(xc, yc, 'k'); 
    % % Place an 'X' at specified angle(s)
    % for theta = angles_for_x
    %     xX = r*cosd(theta);
    %     yX = r*sind(theta);
    %     plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
    % end
end
xX = 2*cosd(0); yX = 2*sind(0);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 2*cosd(90); yX = 2*sind(90);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 2*cosd(180); yX = 2*sind(180);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 2*cosd(270); yX = 2*sind(270);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 0*cosd(0); yX = 0*sind(0);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);

% Dashed lines
xline(0, 'r--','LineWidth',1.5);  
xline(0, 'g--','LineWidth',1.5);
xlim([-5 5]); ylim([-5 5]);
xlabel('X'); ylabel('Y');
axis square
set(gca,'Box','on');

% -- Bottom row (b)
subplot(2,4,6);
hold on;
mu_b = 0;      % shift mean
sigma_b = 1; % narrower or broader as needed
yvals_b = normpdf(xvals, mu_b, sigma_b);
plot(xvals, yvals_b, 'k', 'LineWidth',1.5);
xline(0,'r--','LineWidth',1.5);
xline(mu_b,'g--','LineWidth',1.5);
title('(b) Distribution');
xlabel('x'); ylabel('PDF');
xlim([-5 5]);
axis square
grid on;

%------------------------------------------------
% Subplot (c)
%------------------------------------------------
% -- Top row (c)
subplot(2,4,3); 
hold on; axis equal;
title('(c)');
% % Maybe shift circles far left
% centerC = [-3, 0];
% for r = radii
%     [xc, yc] = circle(centerC(1), centerC(2), r, nCirclePoints);
%     plot(xc, yc, 'k');
%     for theta = angles_for_x
%         xX = centerC(1) + r*cosd(theta);
%         yX = centerC(2) + r*sind(theta);
%         plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
%     end
% end
for r = radii
    [xc, yc] = circle(0,0,r,nCirclePoints);
    plot(xc, yc, 'k'); 
    % % Place an 'X' at specified angle(s)
    % for theta = angles_for_x
    %     xX = r*cosd(theta);
    %     yX = r*sind(theta);
    %     plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
    % end
end
xX = 4.5*cosd(100); yX = 4.5*sind(100);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 4.5*cosd(102); yX = 4.5*sind(102);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 4.5*cosd(104); yX = 4.5*sind(104);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 4.5*cosd(106); yX = 4.5*sind(106);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 4.5*cosd(108); yX = 4.5*sind(108);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);

xline(0,'r--','LineWidth',1.5);
xline(-1.2,'g--','LineWidth',1.5);
xlim([-5 5]); ylim([-5 5]);
xlabel('X'); ylabel('Y');
axis square
set(gca,'Box','on');

% -- Bottom row (c)
subplot(2,4,7);
hold on;
mu_c = -1.2;    
sigma_c = 0.2;  
yvals_c = normpdf(xvals, mu_c, sigma_c);
plot(xvals, yvals_c, 'k','LineWidth',1.5);
xline(0,'r--','LineWidth',1.5);
xline(mu_c,'g--','LineWidth',1.5);
title('(c) Distribution');
xlabel('x'); ylabel('PDF');
xlim([-5 5]);
axis square
grid on;

%------------------------------------------------
% Subplot (d)
%------------------------------------------------
% -- Top row (d)
subplot(2,4,4); 
hold on; axis equal;
title('(d)');
% Circles all centered at (0,0) again
for r = radii
    [xc, yc] = circle(0,0,r,nCirclePoints);
    plot(xc, yc, 'k');
    % for theta = angles_for_x
    %     xX = r*cosd(theta);
    %     yX = r*sind(theta);
    %     plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
    % end
end

xX = 0.5*cosd(0); yX = 0.5*sind(0);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 0.5*cosd(90); yX = 0.5*sind(90);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 0.5*cosd(180); yX = 0.5*sind(180);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 0.5*cosd(270); yX = 0.5*sind(270);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);
xX = 0*cosd(0); yX = 0*sind(0);
plot(xX, yX, 'kx', 'MarkerSize',10, 'LineWidth',2);

% Both lines coincide (as if red & green at x=0)
xline(0, 'r--','LineWidth',1.5);
xline(0, 'g--','LineWidth',1.5);
xlim([-5 5]); ylim([-5 5]);
xlabel('X'); ylabel('Y');
axis square
set(gca,'Box','on');

% -- Bottom row (d)
subplot(2,4,8);
hold on;
% Very narrow distribution around 0
mu_d = 0;
sigma_d = 0.2;
yvals_d = normpdf(xvals, mu_d, sigma_d);
plot(xvals, yvals_d, 'k','LineWidth',1.5);
xline(0,'r--','LineWidth',1.5);
xline(0,'g--','LineWidth',1.5);
title('(d) Distribution');
xlabel('x'); ylabel('PDF');
xlim([-5 5]);
axis square
grid on;

% save_all_figs_OPTION('results/accuracy_vs_precision','png',1)

% --------------------------------------------------
% Helper function: returns x,y points of a circle
% centered at (x0, y0) with radius r
% --------------------------------------------------
function [xC, yC] = circle(x0, y0, r, N)
    if nargin<4
        N = 200; % default resolution
    end
    theta = linspace(0,2*pi,N);
    xC = x0 + r*cos(theta);
    yC = y0 + r*sin(theta);
end

File: Books/A_Comprehensive_Approach_to_ML_and_AI/R/accuracy_vs_precision_tutorial.R
# accuracy_vs_precision_tutorial.R
# This tutorial visualizes accuracy versus precision concepts by plotting
# concentric circles (with markers) and corresponding example distributions.
# Each subplot illustrates a different configuration.

# Clear workspace and close graphics devices
rm(list = ls())
graphics.off()

# Set up common parameters
radii <- c(1, 2, 3, 4, 5)      # Radii of concentric circles
nCirclePoints <- 200          # Number of points to plot each circle

# Define a helper function to compute circle coordinates
circle <- function(x0, y0, r, N = 200) {
  theta <- seq(0, 2 * pi, length.out = N)
  xC <- x0 + r * cos(theta)
  yC <- y0 + r * sin(theta)
  list(x = xC, y = yC)
}

# Set up a 2x4 grid for subplots
par(mfrow = c(2, 4), mar = c(4, 4, 2, 1))

# ---- Subplot (a): Top Row ----
# Plot concentric circles centered at (0,0) with several 'X' markers.
plot(NA, xlim = c(-5, 5), ylim = c(-5, 5), xlab = "X", ylab = "Y",
     main = "(a)", asp = 1)
for (r in radii) {
  circ <- circle(0, 0, r, nCirclePoints)
  lines(circ$x, circ$y, col = "black")
}
# Manually place 'X' markers at specified angles and radii
points(1.5 * cos(0 * pi/180), 1.5 * sin(0 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(4.5 * cos(80 * pi/180), 4.5 * sin(80 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(4.5 * cos(150 * pi/180), 4.5 * sin(150 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(2.5 * cos(200 * pi/180), 2.5 * sin(200 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(4.5 * cos(300 * pi/180), 4.5 * sin(300 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
# Add dashed vertical lines
abline(v = 0, col = "red", lty = 2, lwd = 1.5)
abline(v = 0.75, col = "green", lty = 2, lwd = 1.5)
box()

# ---- Subplot (a): Bottom Row ----
# Plot an example distribution (normal PDF with mean=0.75, sd=2)
subplot_a <- function() {
  xvals <- seq(-10, 10, length.out = 200)
  mu_a <- 0.75
  sigma_a <- 2
  yvals_a <- dnorm(xvals, mean = mu_a, sd = sigma_a)
  plot(xvals, yvals_a, type = "l", lwd = 1.5, col = "black",
       xlab = "x", ylab = "PDF", main = "(a) Distribution", xlim = c(-5, 5), asp = 1)
  abline(v = 0, col = "red", lty = 2, lwd = 1.5)
  abline(v = mu_a, col = "green", lty = 2, lwd = 1.5)
  grid()
}
plot.new()
par(mfrow = c(2, 4))
subplot_a()  # Will be drawn in the 5th subplot

# ---- Subplot (b): Top Row ----
# Plot circles (again) with 'X' markers at a circle of radius 2 at angles 0, 90, 180, 270, and at center.
plot(NA, xlim = c(-5, 5), ylim = c(-5, 5), xlab = "X", ylab = "Y", main = "(b)", asp = 1)
for (r in radii) {
  circ <- circle(0, 0, r, nCirclePoints)
  lines(circ$x, circ$y, col = "black")
}
points(2 * cos(0 * pi/180), 2 * sin(0 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(2 * cos(90 * pi/180), 2 * sin(90 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(2 * cos(180 * pi/180), 2 * sin(180 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(2 * cos(270 * pi/180), 2 * sin(270 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(0, 0, pch = 4, col = "black", cex = 2, lwd = 2)
abline(v = 0, col = "red", lty = 2, lwd = 1.5)
abline(v = 0, col = "green", lty = 2, lwd = 1.5)
box()

# ---- Subplot (b): Bottom Row ----
# Plot normal PDF with mean 0, sd 1
plot(xvals, dnorm(xvals, mean = 0, sd = 1), type = "l", lwd = 1.5, col = "black",
     xlab = "x", ylab = "PDF", main = "(b) Distribution", xlim = c(-5, 5), asp = 1)
abline(v = 0, col = "red", lty = 2, lwd = 1.5)
abline(v = 0, col = "green", lty = 2, lwd = 1.5)
grid()

# ---- Subplot (c): Top Row ----
# Plot circles with 'X' markers at angles between 100 and 108 degrees on circle of radius 4.5.
plot(NA, xlim = c(-5, 5), ylim = c(-5, 5), xlab = "X", ylab = "Y", main = "(c)", asp = 1)
for (r in radii) {
  circ <- circle(0, 0, r, nCirclePoints)
  lines(circ$x, circ$y, col = "black")
}
for (theta in seq(100, 108, by = 2)) {
  points(4.5 * cos(theta * pi/180), 4.5 * sin(theta * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
}
abline(v = 0, col = "red", lty = 2, lwd = 1.5)
abline(v = -1.2, col = "green", lty = 2, lwd = 1.5)
box()

# ---- Subplot (c): Bottom Row ----
# Plot normal PDF with mean -1.2 and sd 0.2
plot(xvals, dnorm(xvals, mean = -1.2, sd = 0.2), type = "l", lwd = 1.5, col = "black",
     xlab = "x", ylab = "PDF", main = "(c) Distribution", xlim = c(-5, 5), asp = 1)
abline(v = 0, col = "red", lty = 2, lwd = 1.5)
abline(v = -1.2, col = "green", lty = 2, lwd = 1.5)
grid()

# ---- Subplot (d): Top Row ----
# Plot circles with small 'X' markers on a circle of radius 0.5
plot(NA, xlim = c(-5, 5), ylim = c(-5, 5), xlab = "X", ylab = "Y", main = "(d)", asp = 1)
for (r in radii) {
  circ <- circle(0, 0, r, nCirclePoints)
  lines(circ$x, circ$y, col = "black")
}
points(0.5 * cos(0 * pi/180), 0.5 * sin(0 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(0.5 * cos(90 * pi/180), 0.5 * sin(90 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(0.5 * cos(180 * pi/180), 0.5 * sin(180 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(0.5 * cos(270 * pi/180), 0.5 * sin(270 * pi/180), pch = 4, col = "black", cex = 2, lwd = 2)
points(0, 0, pch = 4, col = "black", cex = 2, lwd = 2)
abline(v = 0, col = "red", lty = 2, lwd = 1.5)
abline(v = 0, col = "green", lty = 2, lwd = 1.5)
box()

# ---- Subplot (d): Bottom Row ----
# Plot normal PDF with mean 0 and sd 0.2
plot(xvals, dnorm(xvals, mean = 0, sd = 0.2), type = "l", lwd = 1.5, col = "black",
     xlab = "x", ylab = "PDF", main = "(d) Distribution", xlim = c(-5, 5), asp = 1)
abline(v = 0, col = "red", lty = 2, lwd = 1.5)
abline(v = 0, col = "green", lty = 2, lwd = 1.5)
grid()

# (Optional) To save the figure, use the png() function:
# png("accuracy_vs_precision.png", width = 1200, height = 600)
# ... [plotting code] ...
# dev.off()
